### PR TITLE
Made code more intutive

### PR DESCRIPTION
**Pull request by kalgoop**
## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Refactoring (improve a current implementation without adding a new feature or fixing a bug)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Build (changes that affect the build system)
- [ ] Docs (documentation only changes)
- [ ] Test (adding missing tests or fixing existing tests)
- [ ] Other... Please describe:

## Checklist

- [ ] I have read the [CONTRIBUTING](https://docs.ark.io/guidebook/contribution-guidelines/contributing.html) documentation
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: src/helpers/helpers.h [hunk 1]
```diff
@@ -47,7 +47,7 @@
             input /= base;
             (c < 10) ? c += '0' : c += 'A' - 10;
             result = c + result;
-        } while (input);
+        } while (input != 0);
```
Comment (ciband) [inline src/helpers/helpers.h]: Given that this code is dealing with characters, I would rather it say while (input != '\0');.

That being said, does this function even need to exist?  Would the above template cover the uint64_t conversion to string?  I think this is worth testing.

**Comment by sleepdefic1t** [comment]
uin64_t (balances, supply, etc) are too large a container for the `String()`-type on Arduino.
Using the templated method on large integers consistently results in undefined behavior.

I’d be in favor of a unified method for handling this.

Extra kudos if the solution drops the Arduino String construction in favor of string.

Best would be a single method that supports all platforms.

**Comment by ciband** [comment]
Ah, didn't catch the Arduino string.  You can do this:

`
#include <sstream>

template <typename T>
std::string toString(T val) {
  std::ostringstream ss;
  ss << val;
  return ss.str();
}
`

**Comment by ciband** [comment]
ostringstream still uses heap but it is freed quickly.  This pattern can be further expanded with template specialization to provide more efficient conversions for T: integer, T: real using snprintf using local fixed buffers and not doing the dynamic memory allocation.

I thought some of the IoT boards provide the std::to_string functions which essentially do this.

This could also be solved by the pattern used in ark-iot/ark-cpp where we provide a limited implementation of the STL for platforms that don't have it.  Then the code can be written to the C++ STL standard.  That would be a larger change.

**Comment by sleepdefic1t** [comment]
`std::to_string` isn’t available on Arduino because doesn’t include the STL.

The consensus is that it’s too expensive for most IoT platforms.

I’d be interested in how the sstream solution holds up. I think a unified method that maintains IoT support would be preferred over including the STL.

**Comment by ciband** [comment]
The thing is we use the STL on IoT.  ESP32 and ESP8266 support it (some more than others).  The STL is not expensive.  Some things in it are.  I would wager that more platforms will support it over time, espeically as C++11/17/21/xx become more ubiquitous.  

In the current code base we are providing <array> and <memory> implementations when not provided.  They are not expensive in any way.  They just provide additional convenience and safety.

For the purpose of this change, I would definitely leave anything std::to_string out of it and just focus on the potential removal of the function in question.  

We can take the other discussion offline if you like.

**Comment by sleepdefic1t** [comment]
Thems fighting words.
I’ll see you next to the monkey bars after school 💪

32 & 8266 support STL,
the Arduino IDE does not include it.
The compiler flag for c++11 *can* be added with PlatformIO.
Many Arduino boards aren’t as powerful as the Espressif chips though too.

There don’t seem to be many cases where we absolutely need it.
So why include it if we don’t have to?

**Comment by ciband** [comment]
I would submit we use it when ever possible due to convenience.  Without the STL, C++ is just fancy C.

We are already using it on 8266 and 32 and those products work in the Arduino IDE.  The STL is at the board support level, not within the IDE like Arduino or PlatformIO.

We only include what we need from the STL.  It is not all or nothing.  We use it a bunch already (e.g. std::string, std::vector, std::unique_ptr).

**Comment by sleepdefic1t** [comment]
Arduino basically is fancy C,
but is closest to C++ if anything else.
It uses avr-libc.
> http://www.nongnu.org/avr-libc/user-manual/FAQ.html#faq_cplusplus

You can also see this by trying to use `to_string`.
You’ll get a `’to_string’ is not a member of std.` compiling error.

Vectors, strings, and ptr are used,
but it’s not the full c++11 STL, or we wouldn’t need to include it.

The full STL not being included is why it’s not officially considered C++.
And the reason is that while it’s perfect for desktop environments, it’s too big for microcontrollers.
Not that it won’t fit, but that it’s too lax with memory handling to be consistent and stable on IoT.

**Comment by ciband** [comment]
I can agree with some of that.

For to_string, a lot of that is just compilers and libraries lagging behind the bleeding edge.  
The implementation of to string is not inefficient for most things like integers.  Under the hood it will just be a snprintf, which is C and reasonably efficient.  to_string is just nicer.

I think we can have our cake and eat it took.  For our current platforms, this is not a concern.  I think we should look at the problem on a case by case basis as we bring new boards online.

My vote is that we don't solve a problem until we have it and we don't hamstring development by limiting tools when we don't have an explicit reason to.

We leverage everything on every platform we can.  We provide implementations for stuff we want and don't have that is easy and makes sense like std::array and std::unique_ptr.

**Comment by sleepdefic1t** [comment]
Given the above and the following, I see your point:
https://github.com/ArkEcosystem/cpp-crypto/pull/11

I’d be in favor of adding *arduino/STL* to Client & Crypto,
and dropping the `toString` helper methods so long as it maintains consistent support (no reason it shouldn’t).

**Comment by sleepdefic1t** [comment]
We should probably close this out.

OP issue of `while...` isn’t necessary,
and these helper methods will more than likely be dropped.

**Comment by kalgoop** [comment]
Closing this PR, as the former issue isn't much necessary.
If you feel those functions aren't necessary,.. Open a new PR/issue.

**Pull request closed without merging**